crates/extensions/tedge-p11-server/src/pkcs11/mod.rs

```diff
@@ -134,6 +193,48 @@ impl TedgeP11Service for Cryptoki {
         let session = self.open_session_ro(&params)?;
         session.get_public_key_pem()
     }
+
+    fn get_tokens_uris(&self) -> anyhow::Result<Vec<String>> {
+        // refresh the slots before getting slot list
+        let _ = self.reinit();
+        let cryptoki = self.context.lock().unwrap();
+        let slots = cryptoki.get_slots_with_initialized_token().unwrap();
+        let uris = slots
+            .into_iter()
+            .filter_map(|slot| {
+                cryptoki
+                    .get_token_info(slot)
+                    .inspect_err(|e| error!(?e, ?slot, "Failed to get_token_info for slot"))
+                    .map(|t| export_session_uri(&t))
+                    .ok()
+            })
+            .collect();
+
+        Ok(uris)
+    }
+
+    fn create_key(&self, uri: &str, params: CreateKeyParams) -> anyhow::Result<CreateKeyResponse> {
+        let session_params = SessionParams {
+            uri: Some(uri.to_string()),
+            // TODO: do we want to use client-provided (device.key_pin) PIN in create-key-pkcs11 (we can still use
+            // extract PIN from URI)? Options are:
```

Comment:
Is this planned for this PR? If yes, what is the retained approach? The 1st (let the user passes the pin in the URI) and 4th (add a --pin flag that also updates the config) sound reasonable.